thrust/system/hip/hipstdpar/impl/batch.hpp

```diff
@@ -0,0 +1,107 @@
+/*
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*! \file thrust/system/hip/hipstdpar/include/batch.hpp
+ *  \brief <tt>Batch operations</tt> implementation detail header for HIPSTDPAR.
+ */
+
+#pragma once
```

Comment:
Doing the split is fine, since it aligns with Thrust convention, I guess; however, IMHO it'd be better if the master header (this), which is the object of implicit inclusion / is part of the public interface, would do nothing but include the split headers, as opposed to having any actual functionality. We could also probably guard the `interpose_allocations.hpp` one with its respective macros. Finally, we probably want to error on the impl headers if they're directly included without `__HIPSTDPAR__` being defined, as that would indicate user error.

Reply:
> Do you mean that we should further split the master headers (the ones named after the C++ Standard Library categories for algorithms)?

Not quite, what I meant is that the master header would pretty much be a list of `#include`s for the impl headers, with no declarations / definitions or, really, any substantive code living there. In essence, the most we'd do here is check for environmental macros / possibly declare some future config macros, if necessary. Does that make sense?

> What macros do you refer to? IINM the `interpose_allocations.hpp` [is already guarded](https://github.com/ROCm/rocThrust/blob/8aff938930142937a527c1d54bc202ba57c26183/thrust/system/hip/hipstdpar/impl/interpose_allocations.hpp#L21-L22) by the `__HIPSTDPAR__` and `__HIPSTDPAR_INTERPOSE_ALLOC__` macros (as it was before splitting `hipstdpar_lib.hpp`).

What I was suggesting as a possible alternative is to check for `__HIPSTDPAR_INTERPOSE_ALLOC__` in the master `hipstdpar_lib.hpp` header and do conditional include of the impl header in there, rather than guarding within it. This ties both to the point above re: what to do with the master header and to the one below about considering error-ing out.
 
> And for this, I don't fully understand why would we need to error here. Originally there was no error handling in case that macro was not defined in the `hipstdpar_lib.hpp`, so I'm not sure why it would be done for the impl headers. Also, as I had understood it, the forwarding header was not intended to be included directly by users, but implicitly included by the compiler when specifying the `--hipstdpar` option (which also triggers the definition of the `__HIPSTDPAR__` macro). So if this is true (please correct me if not), then it wouldn't make sense to check for the definition of this macro IMHO.

Whilst the vote of confidence is highly appreciated, I'd not take the original header to necessarily be complete / the best way to do things:) To wit, whilst the forwarding header was / is not meant to be included by users, that does not mean users won't do it (accidentally / intentionally). Error-ing out if the macro is not defined i.e. compilation for `hipstdpar` was not desired would guard against accidents, and will at least force intentional abusers to manually define the macro, which'd help with eventual debugging. I suggested it because I think it might be a time saver / inform users about things potentially going awry. Ultimately, it's up to you if you want to go in this direction.

Reply:
> Not quite, what I meant is that the master header would pretty much be a list of #includes for the impl headers, with no declarations / definitions or, really, any substantive code living there. In essence, the most we'd do here is check for environmental macros / possibly declare some future config macros, if necessary. Does that make sense?

Oh alright, I was confused because the comment was made on `impl/batch.hpp`. But then, the master header `hipstdpar_lib.hpp` is already just a macro definition check (`__HIPSTDPAR__`) and a bunch of includes (for the `impl/` headers)

> What I was suggesting as a possible alternative is to check for __HIPSTDPAR_INTERPOSE_ALLOC__ in the master hipstdpar_lib.hpp header and do conditional include of the impl header in there, rather than guarding within it. This ties both to the point above re: what to do with the master header and to the one below about considering error-ing out.

> Whilst the vote of confidence is highly appreciated, I'd not take the original header to necessarily be complete / the best way to do things:) To wit, whilst the forwarding header was / is not meant to be included by users, that does not mean users won't do it (accidentally / intentionally). Error-ing out if the macro is not defined i.e. compilation for hipstdpar was not desired would guard against accidents, and will at least force intentional abusers to manually define the macro, which'd help with eventual debugging. I suggested it because I think it might be a time saver / inform users about things potentially going awry. Ultimately, it's up to you if you want to go in this direction.

Regarding those two points, the final logic should be then:
- Check in `hipstdpar_lib.hpp` if `__HIPSTDPAR__` is defined and include the `impl` headers if so, except `interpose_allocations.hpp` which only will be included if `__HIPSTDPAR_INTERPOSE_ALLOC__` is also defined.
- Inside each implementation header, check if `__HIPSTDPAR__` is defined and err if it's not.
- Additionally, in `interpose_allocations.hpp`, check if `__HIPSTDPAR_INTERPOSE_ALLOC__` is defined and err if not.

If that sounds good I'll add a commit with the changes

Reply:
This sounds great, thanks and apologies for adding work.

Reply:
Also added the err logic discussed before